pubsys: add `validate-repo`, `check-repo-expirations`, `refresh-repo` subcommands

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
Fixes https://github.com/bottlerocket-os/bottlerocket/issues/1117


**Description of changes:**
```
Author: Erikson Tung <etung@amazon.com>
Date:   Mon Sep 21 18:06:26 2020 -0700

    pubsys: add `validate-repo` for validating TUF repositories
    
    Adds a new pubsys subcommand `validate-repo` for validating a set of TUF
    repositories.
    
    Adds a new cargo make task `validate-repo` that invokes this new subcommand.

```

```
Author: Erikson Tung <etung@amazon.com>
Date:   Wed Sep 23 14:52:07 2020 -0700

    pubsys: add `check-repo-expirations` to check metadata expirations
    
    Adds new pubsys subcommand `check-repo-expirations` that checks for
    repository metadata expirations within a specified timeframe
    
    Adds a new Makefile.toml task that invokes `check-repo-expirations`
```

```
Author: Erikson Tung <etung@amazon.com>
Date:   Thu Sep 24 15:39:27 2020 -0700

    pubsys: add `refresh-repo` to refresh and resign TUF repositories
    
    Adds a new subcommand `refresh-repo` for refreshing and resigning
    non-root metadata files of TUF repositories.
    
    Add new Makefile.toml task `refresh-repo` for refreshing the expiration
    dates of TUF repositories' metadata files
```

**Testing done:**

**Validating repositories**
Normal validation run via `cargo make validate-repo` where 2020-07-07/aws-ecs-1/aarch64 is successfully loaded and its targets successfully retrieved.
<details>

```
$ cargo make -e PUBLISH_REPO="2020-07-07" -e BUILDSYS_VARIANT="aws-ecs-1" -e BUILDSYS_ARCH="aarch64" validate-repo
[cargo-make] INFO - cargo make 0.32.5
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: validate-repo
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup-without-key
00:42:54 [INFO] Found infra config at path: /home/etung/thar/Infra.toml
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: validate-repo
00:43:47 [INFO] Using infra config from path: /home/etung/thar/Infra.toml
00:43:48 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-ecs-1/aarch64
00:43:48 [INFO] Downloading 100% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-ecs-1/aarch64
00:43:48 [INFO] Downloading target: bottlerocket-aws-ecs-1-aarch64-1.0.0-b0e2bc22-boot.ext4.lz4
00:43:49 [INFO] Downloading target: bottlerocket-aws-ecs-1-aarch64-1.0.1-2a181156-root.verity.lz4
00:43:49 [INFO] Downloading target: migrate_v0.4.1_add-version-lock-ignore-waves.lz4
00:43:49 [INFO] Downloading target: migrate_v1.0.0_ecr-helper-admin.lz4
00:43:49 [INFO] Downloading target: migrate_v1.0.0_ecr-helper-control.lz4
00:43:49 [INFO] Downloading target: bottlerocket-aws-ecs-1-aarch64-1.0.0-b0e2bc22-root.ext4.lz4
00:43:58 [INFO] Downloading target: migrate_v0.5.0_admin-container-v0-5-2.lz4
00:43:58 [INFO] Downloading target: migrate_v0.5.0_control-container-v0-4-1.lz4
00:43:59 [INFO] Downloading target: bottlerocket-aws-ecs-1-aarch64-1.0.1-2a181156-boot.ext4.lz4
00:44:00 [INFO] Downloading target: manifest.json
00:44:00 [INFO] Downloading target: bottlerocket-aws-ecs-1-aarch64-1.0.0-b0e2bc22-root.verity.lz4
00:44:00 [INFO] Downloading target: migrate_v0.4.1_pivot-repo-2020-07-07.lz4
00:44:00 [INFO] Downloading target: bottlerocket-aws-ecs-1-aarch64-1.0.1-2a181156-root.ext4.lz4
00:44:09 [INFO] Downloading target: migrate_v0.5.0_add-cluster-domain.lz4
00:44:09 [INFO] Downloading target: migrate_v0.3.2_admin-container-v0-5-0.lz4
[cargo-make] INFO - Build Done in 84 seconds.
```

</details>

Test validation run for 2020-07-07 repositories where I disable internet access halfway through to test error handling. 

<details>

```
$ cargo make -e PUBLISH_REPO="2020-07-07" -e BUILDSYS_VARIANT="aws-k8s-1.17" -e BUILDSYS_ARCH="x86_64" validate-repo
[cargo-make] INFO - cargo make 0.32.7
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: validate-repo
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup
17:05:16 [INFO] Found infra config at path: /Infra.toml
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: validate-repo
17:05:17 [INFO] Using infra config from path: /Infra.toml
17:05:20 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/x86_64
17:05:20 [INFO] Attempting to download 100% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/x86_64 for validation purposes
17:05:20 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-0.5.0-e0ddf1b-boot.ext4.lz4
17:05:20 [INFO] Downloading target: migrate_v0.4.1_pivot-repo-2020-07-07.lz4
17:05:21 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-0.5.0-e0ddf1b-root.ext4.lz4
17:05:21 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.1-2a181156-root.verity.lz4
17:05:22 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.1-2a181156-boot.ext4.lz4
17:05:22 [INFO] Downloading target: migrate_v0.5.0_admin-container-v0-5-2.lz4
17:05:23 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.0-b0e2bc22-root.verity.lz4
17:05:23 [INFO] Downloading target: migrate_v0.4.1_add-version-lock-ignore-waves.lz4
17:05:23 [INFO] Downloading target: migrate_v0.3.2_admin-container-v0-5-0.lz4
17:05:24 [INFO] Downloading target: migrate_v1.0.0_ecr-helper-control.lz4
17:05:24 [INFO] Downloading target: migrate_v0.5.0_control-container-v0-4-1.lz4
Failed to validate repository: Failed to read target 'manifest.json' from repo: Failed to fetch https://updates.bottlerocket.aws/targets/004c373cb0899ed114a94692e245c4e45eeaa74155470de74b9a234c14920800.manifest.json: Could not fetch repo at 'https://updates.bottlerocket.aws/targets/004c373cb0899ed114a94692e245c4e45eeaa74155470de74b9a234c14920800.manifest.json': Failed to fetch 'https://updates.bottlerocket.aws/targets/004c373cb0899ed114a94692e245c4e45eeaa74155470de74b9a234c14920800.manifest.json' after 4 tries, final error: error sending request for url (https://updates.bottlerocket.aws/targets/004c373cb0899ed114a94692e245c4e45eeaa74155470de74b9a234c14920800.manifest.json): operation timed out
[cargo-make] ERROR - Error while executing command, exit code: 1
[cargo-make] WARN - Build Failed.


```

</details>

**Checking for repository non-root metadata expirations**
Successfully ran the `check-repo-expirations` on 2020-02-02 prefixed Bottlerocket TUF repositories.
<details>

```
$ cargo make -e PUBLISH_REPO="2020-02-02" -e BUILDSYS_VARIANT="aws-k8s-1.17" -e BUILDSYS_ARCH="x86_64" -e check-repo-expirations
[cargo-make] INFO - cargo make 0.32.5
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: check-repo-expirations
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup-without-key
21:05:45 [INFO] Found infra config at path: /home//Infra.toml
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: check-repo-expirations
21:05:45 [INFO] Using infra config from path: /home/Infra.toml
21:05:49 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-02-02/aws-k8s-1.17/x86_64
21:05:49 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
21:05:49 [INFO] Snapshot expiration:	2020-10-21 17:52:25.534814597 UTC
21:05:49 [INFO] Targets expiration:	2020-10-21 17:52:25.534815167 UTC
21:05:49 [INFO] Timestamp expiration:	2020-10-14 17:52:25.534815678 UTC
21:05:49 [INFO] Looking for metadata expirations happening from now to 2020-10-11 21:05:45.995390949 UTC
[cargo-make] INFO - Build Done in 5 seconds.
```

</details>

Specifying `--upcoming-expirations-in 10` will cause pubsys to list timestamp metadata files in upcoming expirations, which is expected.
<details>

```
$ cargo make -e PUBLISH_REPO="2020-02-02" -e BUILDSYS_VARIANT="aws-k8s-1.17" -e BUILDSYS_ARCH="x86_64" -e EXPIRING_WITHIN='10 days' check-repo-expirations
[cargo-make] INFO - cargo make 0.32.5
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: check-repo-expirations
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup-without-key
21:07:35 [INFO] Found infra config at path: /home//Infra.toml
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: check-repo-expirations
21:07:36 [INFO] Using infra config from path: /home//Infra.toml
21:07:38 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-02-02/aws-k8s-1.17/x86_64
21:07:38 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
21:07:38 [INFO] Snapshot expiration:	2020-10-21 17:52:25.534814597 UTC
21:07:38 [INFO] Targets expiration:	2020-10-21 17:52:25.534815167 UTC
21:07:38 [INFO] Timestamp expiration:	2020-10-14 17:52:25.534815678 UTC
21:07:38 [INFO] Looking for metadata expirations happening from now to 2020-10-18 21:07:36.089668869 UTC
21:07:38 [WARN] Repo 'https://updates.bottlerocket.aws/2020-02-02/aws-k8s-1.17/x86_64': 'timestamp' expiring between now and 2020-10-18 21:07:36.089668869 UTC on 2020-10-14 17:52:25.534815678 UTC
Found expiring/expired metadata in 'https://updates.bottlerocket.aws/2020-02-02/aws-k8s-1.17/x86_64'
[cargo-make] ERROR - Error while executing command, exit code: 1
[cargo-make] WARN - Build Failed.
```

</details>

**Refreshing and resigning repository non-root metadata files**
Successfully refreshed and resigned metadata for a test TUF repository using a given expiration policy

<details>

```
$ cargo make -e PUBLISH_REPO="etung" -e BUILDSYS_VARIANT="aws-k8s-1.17" -e BUILDSYS_ARCH="aarch64" -e UNSAFE_REFRESH=true refresh-repo
[cargo-make] INFO - cargo make 0.32.5
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: refresh-repo
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup
20:31:50 [INFO] Found infra config at path: /home//Infra.toml
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: refresh-repo
20:32:42 [INFO] Using infra config from path: /home/Infra.toml
20:32:42 [INFO] Using repo expiration policy from path: /home/tools/pubsys/policies/repo-expiration/2w-2w-1w.toml
20:32:45 [INFO] Loaded TUF repo: https://d31e403t2udbbh.cloudfront.net/aws-k8s-1.17/aarch64
20:32:45 [INFO] Setting non-root metadata expiration times:
	snapshot:  2020-10-22 20:32:45.590842865 UTC
	targets:   2020-10-22 20:32:45.590842865 UTC
	timestamp: 2020-10-15 20:32:45.590842865 UTC
20:32:45 [INFO] Writing repo metadata to: /home/build/repos/etung/bottlerocket-1.0.2-5119490c/aws-k8s-1.17/aarch64
[cargo-make] INFO - Build Done in 60 seconds.
```

</details>

The TUF repository passes validation checks after uploading the refreshed & resigned metadata files

<details>

```
$ cargo make -e PUBLISH_REPO="etung" -e BUILDSYS_VARIANT="aws-k8s-1.17" -e BUILDSYS_ARCH="x86_64" validate-repo
[cargo-make] INFO - cargo make 0.32.5
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: validate-repo
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup-without-key
16:57:33 [INFO] Found infra config at path: /home/etung/thar/Infra.toml
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: validate-repo
16:57:34 [INFO] Using infra config from path: /home/etung/thar/Infra.toml
16:57:34 [INFO] Loaded TUF repo: https://asdasdasdasdasd.cloudfront.net/aws-k8s-1.17/x86_64
16:57:34 [INFO] Downloading 100% of listed targets from https://asdasdasdasd.cloudfront.net/aws-k8s-1.17/x86_64
16:57:34 [INFO] Downloading target: 2.txt
16:57:34 [INFO] Downloading target: 1.txt
[cargo-make] INFO - Build Done in 2 seconds.

```

</details>

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten commented:
Some minor touches in the push above.

tjkirch commented on tools/pubsys/Infra.toml.example:
```diff
@@ -27,6 +27,18 @@ signing_keys = { default = { file = { path = "/home/user/key.pem" } } }
 # metadata_base_url = "https://example.com/metadata/"
 # targets_url = "https://example.com/targets/"
 
+# Existing arches and variants under the specific `metadata_base_url` (prefix) can be specified here.
```
```suggestion
# Arches and variants you publish to this repo.  This allows you to use the repo validation targets in the makefile.
```

tjkirch commented on tools/pubsys/Infra.toml.example:
```diff
@@ -27,6 +27,18 @@ signing_keys = { default = { file = { path = "/home/user/key.pem" } } }
+#    "aws-ecs-1",
+# ]
+# When validating repositories, if this "repo" is selected, the set of repository urls are composed
+# from its set of arches and variants like so:
```
```suggestion
# Repository URLs are expected to match the structure in the `build/repos` directories:
```

tjkirch commented on tools/pubsys/src/main.rs:
```diff
@@ -151,6 +156,9 @@ mod error {
         Repo { source: crate::repo::Error },
 
+        #[snafu(display("Failed to validate repo(s): {}", source))]
+        ValidateRepo { source: crate::repo::Error },
```
Can we please separate `repo` from `validate_repo` errors?

zmrow commented on tools/pubsys/src/main.rs:
```diff
@@ -151,6 +161,12 @@ mod error {
+        #[snafu(display("Failed to validate repo(s): {}", source))]
+        ValidateRepo { source: crate::repo::Error },
+
+        #[snafu(display("{}", source))]
```
If you don't need to provide any additional context to the user beyond what the underlying error source gives you, you can use `#[snafu(context(false))]` rather than the empty `display()`. [See here](https://docs.rs/snafu/0.6.6/snafu/guide/attributes/index.html#controlling-context).

(I just learned about this yesterday! :smile: )

etungsten replied:
I'm starting to realize this might not be what we want here. I can only use `#[snafu(context(false))]` for one error that has `crate::repo::Error` as the `source`. If I have more than one, then snafu gets confused when trying to convert the underlying `repo::Error` as the top level pubsys `Error`. 

The workaround is to have separate error types for each subcommand `validate_repo`, `check_repo_expiration` and `refresh_repo`. But then I would end up with a bunch of duplicate Errors so I'm reluctant to do that.

```
error[E0119]: conflicting implementations of trait `std::convert::From<repo::error::Error>` for type `error::Error`:
   --> pubsys/src/main.rs:148:21
    |
148 |     #[derive(Debug, Snafu)]
    |                     ^^^^^
    |                     |
    |                     first implementation here
    |                     conflicting implementation for `error::Error`
    |
    = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0119`.
error: could not compile `pubsys`.
```

zmrow replied:
Makes sense - best to reverse it.

zmrow commented on tools/pubsys/src/repo.rs:
```diff
@@ -286,6 +288,71 @@ fn repo_urls<'a>(
+        })?;
+
+    // Make sure all fields are set and then compose the set of metadata URLs to validate
+    if let Some(metadata_base_url) = &repo_config.metadata_base_url {
```
I totally realize this is an unfortunate side effect of having so many `Option`s. :(

To avoid all the extra nesting, let's make sure all the `Options` are `Some()`, then we can just use them.  Having said that, checking if they're all `Some()` isn't super pretty either, but avoiding the extra nesting/indentation is worth it.  My off the cuff suggestion is:
```
// Has the disadvantage that it's not very pretty
If let (Some(foo), Some(bar), ...) = (metadata_base_url, targets_url, ...) {
   <do the thing>
} else {
   error
}
```

zmrow commented on tools/pubsys/src/repo.rs:
```diff
@@ -286,6 +288,71 @@ fn repo_urls<'a>(
+                        .map(|arch| {
+                            variants
+                                .iter()
+                                .map(|variant| {
```
I might be alone here, but I think a few for loops here may read a bit clearer.  I'm a bit lost by the `.fold()` at the end.

etungsten replied:
I agree it is kinda confusing. The `.fold()` accumulates the `collect()`ed `Vec<String>`s from the "inner" iteration into a final set of URLs.
Replaced it with a double for loop for clarity.

etungsten commented:
Addresses @zmrow 's comments.
Tested things and they still work as before.

etungsten commented:
For `check-repo-expirations`, changed `--upcoming-expirations-in` option to `--expirations-between-now-and`, the option now takes a RFC3339 date or short-hand "in X hours/days/weeks" as an argument.

Updated PR testing description.

Push below fixes the Makefile.toml `check-repo-expirations` task

zmrow commented on tools/pubsys/src/repo.rs:
```diff
@@ -286,6 +288,61 @@ fn repo_urls<'a>(
+        })?;
+
+    // Make sure all fields are set and then compose the set of metadata URLs
+    if let (Some(metadata_base_url), Some(targets_url), Some(arches), Some(variants)) = (
```
Is there any reason not to throw an error if one of these isn't set? You can't compose a valid repo URL.

etungsten replied:
I do throw an error in `validate_repo::run()`; if this function returns `Ok(None)` then it means I can't compose any repo URLs of the format `${metadata_base_url}${base_slash}${variant}/${arch}`

etungsten replied:
Oh I guess I could just return a result instead of an option in that case. Silly me

etungsten committed:
pubsys: add `refresh-repo` to refresh and resign TUF repositories

Adds a new subcommand `refresh-repo` for refreshing and resigning
non-root metadata files of TUF repositories.

Add new Makefile.toml task `refresh-repo` for refreshing the expiration
dates of TUF repositories' metadata files

etungsten commented:
Push above does the following:

- Addresses @zmrow 's comment: https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r494642180 
- Reversed changes made for https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r494410061. See comment below for reasons why.
- Add new subcommand `refresh-repo`.

zmrow commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,186 @@
+            }
+            let target_reader = repo.read_target(&target);
+            match target_reader {
+                Err(err) => match err {
+                    tough::error::Error::Transport { .. } => {
+                        error!("Fetch failure on attempt {}", attempt);
+                        if attempt >= max_attempts {
+                            return Some(repo_error::Error::ReadTarget {
+                                target,
+                                source: err,
+                            });
```
The logic from here on down is fairly nested - I'm wondering if it should get split out into a function so it's easier to follow and read.

etungsten replied:
I feel like it would be somewhat more confusing if I split out the stuff in `match target_reader` into it's own function due to having to keep track of the current iteration of the loop for the purpose of implementing retries and determining the proper error to return. 

I can add more comments for clarity. Do you think that'll help?

etungsten replied:
Scratch the above, I think I know what I can do.

zmrow reviewed (commented):
I noticed in the testing you're only poking at `x86_64` repos, is this by design? It looks like the code should handle multiple architectures.

etungsten commented:
> I noticed in the testing you're only poking at `x86_64` repos, is this by design? It looks like the code should handle multiple architectures.

Oh that's because I was mostly testing with the 2020-02-02 repos which only had x86_64 as its only architecture. 

Multiple architecture works.
2020-07-07 repositories:

Check expirations

<details>

```
$ cargo run -- --log-level info --infra-config-path ./Infra.test.toml check-repo-expirations --repo 2020-07-07
16:46:58 [INFO] Using infra config from path: ./Infra.test.toml
16:47:01 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.15/x86_64
16:47:01 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
16:47:01 [INFO] Snapshot expiration:	2020-10-08 21:17:27.091923992 UTC
16:47:01 [INFO] Targets expiration:	2020-10-08 21:17:27.091924651 UTC
16:47:01 [INFO] Timestamp expiration:	2020-10-01 21:17:27.091925099 UTC
16:47:04 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.16/x86_64
16:47:04 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
16:47:04 [INFO] Snapshot expiration:	2020-10-08 21:17:39.876516265 UTC
16:47:04 [INFO] Targets expiration:	2020-10-08 21:17:39.876516866 UTC
16:47:04 [INFO] Timestamp expiration:	2020-10-01 21:17:39.876517309 UTC
16:47:06 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/x86_64
16:47:06 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
16:47:06 [INFO] Snapshot expiration:	2020-10-08 21:17:54.679132007 UTC
16:47:06 [INFO] Targets expiration:	2020-10-08 21:17:54.679132550 UTC
16:47:06 [INFO] Timestamp expiration:	2020-10-01 21:17:54.679133085 UTC
16:47:08 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-07-07/aws-ecs-1/x86_64
16:47:08 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
16:47:08 [INFO] Snapshot expiration:	2020-10-08 21:18:08.656811373 UTC
16:47:08 [INFO] Targets expiration:	2020-10-08 21:18:08.656811921 UTC
16:47:08 [INFO] Timestamp expiration:	2020-10-01 21:18:08.656812376 UTC
16:47:10 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.15/aarch64
16:47:10 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
16:47:10 [INFO] Snapshot expiration:	2020-10-08 21:17:34.016295924 UTC
16:47:10 [INFO] Targets expiration:	2020-10-08 21:17:34.016296533 UTC
16:47:10 [INFO] Timestamp expiration:	2020-10-01 21:17:34.016297033 UTC
16:47:12 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.16/aarch64
16:47:12 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
16:47:12 [INFO] Snapshot expiration:	2020-10-08 21:17:48.771882884 UTC
16:47:12 [INFO] Targets expiration:	2020-10-08 21:17:48.771883458 UTC
16:47:12 [INFO] Timestamp expiration:	2020-10-01 21:17:48.771883912 UTC
16:47:15 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/aarch64
16:47:15 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
16:47:15 [INFO] Snapshot expiration:	2020-10-08 21:18:02.773344554 UTC
16:47:15 [INFO] Targets expiration:	2020-10-08 21:18:02.773345108 UTC
16:47:15 [INFO] Timestamp expiration:	2020-10-01 21:18:02.773345581 UTC
16:47:17 [INFO] Loaded TUF repo:	https://updates.bottlerocket.aws/2020-07-07/aws-ecs-1/aarch64
16:47:17 [INFO] Root expiration:	2021-09-22 00:00:00 UTC
16:47:17 [INFO] Snapshot expiration:	2020-10-08 21:18:12.610542929 UTC
16:47:17 [INFO] Targets expiration:	2020-10-08 21:18:12.610543512 UTC
16:47:17 [INFO] Timestamp expiration:	2020-10-01 21:18:12.610543975 UTC
```

</details>

Validate repos

<details>

```
$ cargo run -- --log-level info --infra-config-path ./Infra.test.toml validate-repo --repo 2020-07-07 --percent-target-files 5
16:48:45 [INFO] Using infra config from path: ./Infra.test.toml
16:48:49 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.15/x86_64
16:48:49 [INFO] Downloading 5% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.15/x86_64
16:48:50 [INFO] Downloading target: migrate_v0.4.1_add-version-lock-ignore-waves.lz4
16:48:51 [INFO] Downloading target: migrate_v0.5.0_add-cluster-domain.lz4
16:48:54 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.16/x86_64
16:48:54 [INFO] Downloading 5% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.16/x86_64
16:48:55 [INFO] Downloading target: bottlerocket-aws-k8s-1.16-x86_64-0.4.1-a29a1450-boot.ext4.lz4
16:48:57 [INFO] Downloading target: bottlerocket-aws-k8s-1.16-x86_64-1.0.0-b0e2bc22-root.verity.lz4
16:49:00 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/x86_64
16:49:00 [INFO] Downloading 5% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/x86_64
16:49:01 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.0-b0e2bc22-boot.ext4.lz4
16:49:03 [INFO] Downloading target: migrate_v0.4.1_add-version-lock-ignore-waves.lz4
16:49:06 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-ecs-1/x86_64
16:49:06 [INFO] Downloading 5% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-ecs-1/x86_64
16:49:07 [INFO] Downloading target: migrate_v0.5.0_admin-container-v0-5-2.lz4
16:49:09 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.15/aarch64
16:49:09 [INFO] Downloading 5% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.15/aarch64
16:49:10 [INFO] Downloading target: bottlerocket-aws-k8s-1.15-aarch64-1.0.1-2a181156-root.verity.lz4
16:49:13 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.16/aarch64
16:49:13 [INFO] Downloading 5% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.16/aarch64
16:49:14 [INFO] Downloading target: migrate_v1.0.0_ecr-helper-control.lz4
16:49:17 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/aarch64
16:49:17 [INFO] Downloading 5% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/aarch64
16:49:18 [INFO] Downloading target: migrate_v0.4.1_add-version-lock-ignore-waves.lz4
16:49:20 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-ecs-1/aarch64
16:49:20 [INFO] Downloading 5% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-ecs-1/aarch64
16:49:21 [INFO] Downloading target: bottlerocket-aws-ecs-1-aarch64-1.0.0-b0e2bc22-root.ext4.lz4
```

</details>

etungsten commented:
Refactored `retrieve_percentage_of_targets` to address https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r495091246.
Addressed some `cargo clippy` items. 

Things still work as before, updated testing description.

tjkirch commented on Makefile.toml:
```diff
@@ -44,6 +44,11 @@ PUBLISH_DATA_VOLUME_SIZE = "20"
+# Percentage of TUF repository targets to sample and retrieve as part of the repo validation process
+VALIDATE_PERCENT_TARGETS = "100"
+# Specifies the timeframe to look for upcoming repository metadata expirations
+EXPIRATIONS_BETWEEN_NOW_AND = "in 3 days"
```
```suggestion
EXPIRING_WITHIN = "3 days"
```

zmrow reviewed (approved):
:beach_umbrella:

etungsten commented:
Going to wait for https://github.com/bottlerocket-os/bottlerocket/pull/1133 to be merged and resolve any conflicts before merging this.

tjkirch commented on tools/pubsys/src/repo/refresh_repo/mod.rs:
```diff
@@ -0,0 +1,214 @@
+        let mut metadata_out_dir = refresh_repo_args.outdir.join(
+            metadata_url
+                .domain()
+                .unwrap_or_else(|| metadata_url.as_str()),
```
How does this work with file URLs?  I don't think "domain" really has a meaning there (if this works, it's probably not the part of the URL someone would expect) and falling back to the entire metadata URL doesn't seem safe because it's almost certainly going to have slashes and break the path structure.

[edit] after rebasing on #1135, I think this will kind of automatically use the repo name instead of the metadata URL, which should be good.

tjkirch commented on Makefile.toml:
```diff
@@ -93,6 +98,9 @@ PUBLISH_REPO_OUTPUT_DIR = "${PUBLISH_REPO_BASE_DIR}/${BUILDSYS_NAME_VERSION}"
 # The default name of registered AMIs; override by setting PUBLISH_AMI_NAME.
 PUBLISH_AMI_NAME_DEFAULT = "${BUILDSYS_NAME}-${BUILDSYS_VARIANT}-${BUILDSYS_ARCH}-v${BUILDSYS_VERSION_IMAGE}-${BUILDSYS_VERSION_BUILD}"
 
+# Repo directories for the refresh and resign non-root metadata task
+REFRESH_REPO_OUTPUT_DIR = "${PUBLISH_REPO_BASE_DIR}/refreshed-repos"
```
I think it'd be better to use the existing PUBLISH_REPO_OUTPUT_DIR, and (like in the `repo` task) fail if that directory exists.  I think people will rarely be doing both of those things in the same checkout, and it's clearer to have one spot for repos.

(After rebasing on #1135, it'll be easier to use the same path, since we can pass in PUBLISH_REPO_OUTPUT_DIR, which now includes the repo name, rather than basing the path on the metadata URL.)

tjkirch commented on tools/pubsys/src/repo.rs:
```diff
@@ -1,6 +1,9 @@
 //! The repo module owns the 'repo' subcommand and controls the process of building a repository.
 
+pub(crate) mod check_expirations;
+pub(crate) mod refresh_repo;
```
We talked about this offline -- I think it's worth seeing how hard it'd be to update `repo.rs` to handle the case of updating a repo without adding any versions/targets, rather than adding a whole new module.  There's a lot of shared process, so it'd be nice not to duplicate it, but if it's hard to combine the two then we can stick with the separate module for now and maybe come back to it later.

(Maybe something like a flattened sub-struct in `RepoArgs` that has the ~5 arguments that get used when adding a version, and if those aren't present, we refresh the repo rather than adding a new version.)

etungsten replied:
I tried to refactor `repo.rs` to accommodate refreshing repo metadata without adding targets or images and I couldn't get it to a state that I'm happy with. The changeset became too big to manage in this PR and would impact the original repo creation functionality. So I'd like to create a separate issue to follow up with this if that's OK.

tjkirch replied:
Sure, we can also just drop it; the commands are different enough.

etungsten commented:
Push above does the following:
- Separate out errors for the newly added subcommands into their own modules (addresses https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r493949541)
- Addresses https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r493873726, https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r493948691, https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r495309207

Still need to rebase onto https://github.com/bottlerocket-os/bottlerocket/pull/1133# and address the remaining comments.

bcressey commented on tools/pubsys/Cargo.toml:
```diff
@@ -16,6 +16,7 @@ indicatif = "0.15.0"
 lazy_static = "1.4"
 log = "0.4"
 parse-datetime = { path = "../../sources/parse-datetime" }
+rand = "0.7.3"
```
Elsewhere we've used this to trim the size of the dependency:
```
rand = { version = "0.7", default-features = false, features = ["std"] }
```

bcressey commented on tools/pubsys/Infra.toml.example:
```diff
@@ -27,6 +27,17 @@ signing_keys = { default = { file = { path = "/home/user/key.pem" } } }
 # metadata_base_url = "https://example.com/metadata/"
 # targets_url = "https://example.com/targets/"
 
+# Arches and variants you publish to this repo.  This allows you to use the repo validation targets in the makefile.
+# arches = [ "x86_64", "aarch64" ]
+# variants = [
+#    "aws-k8s-1.15",
+#    "aws-k8s-1.16",
+#    "aws-k8s-1.17",
+#    "aws-ecs-1",
+# ]
+# Repository URLs are expected to match the structure in the `build/repos` directories:
+#     ${metadata_base_url}/${variant}/${arch}/
```
I'm not excited about adding these to Infra.toml, as it changes the abstraction that this is merely a description of available infrastructure, by overloading it to also include policy on how that infrastructure is or should be used.

For example, it seems reasonable to expect that what's asserted here will be used for all the repo commands, and that `cargo make repo` would block an attempt to build a repo for `aws-dev` if it's not in the list.

Also, I might not want to support all architectures for all variants - `rpi-dev` might only exist for aarch64, even if I used the same endpoints for hosting other variants that did support x86_64.

The idiomatic way to check multiple variants and architectures would be to loop over `cargo make` invocations, modifying `BUILDSYS_VARIANT` and `BUILDSYS_ARCH` as appropriate.

etungsten replied:
What if we create a new table named `[existing-repo.NAME]` to house the arch variant combinations? This could also be a separate config file (Repos.toml?) from Infra.toml but would still be collated when calling pubsys validation subcommands in cargo make.

Something like:
```
[existing-repo.2020-02-02]
// These existing repositories are created using this repository infrastructure from Infra.toml
repo_infra = "repo.2020-02-02"
// The valid arch variant combinations
arches = [ "x86_64", "aarch64" ]
variants = [
    "aws-k8s-1.15",
    "aws-k8s-1.16",
    "aws-k8s-1.17",
    "aws-ecs-1",
]

[existing-repo.raspberry-pi]
repo_infra = "repo.${some_key}"
arches = ["aarch64"]
variants = [ "rpi-dev" ]
```

webern replied:
> The idiomatic way to check multiple variants and architectures would be to loop over `cargo make` invocations, modifying `BUILDSYS_VARIANT` and `BUILDSYS_ARCH` as appropriate.

Your concern seems valid. The issue then becomes, where should the list of variants and arches to loop over be stored? Possibly we are asking Infra.toml to do too much?

tjkirch replied:
> where should the list of variants and arches to loop over be stored?

There was some previous discussion of using a file in the repo bucket itself to enumerate what lives within.  Getting the format right could be tricky, but it would separate concerns.

etungsten commented:
Push above rebases onto develop to include changes from #1133 and fixes conflicts.

The push also addresses https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r496311355, https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r496294642, https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r497182329.


I haven't addressed https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r497185358 yet.

etungsten commented:
Push above is just a rebase on develop

etungsten committed:
pubsys: add `validate-repo` for validating TUF repositories

Adds a new pubsys subcommand `validate-repo` for validating TUF repositories.

Adds a new cargo make task `validate-repo` that invokes this new subcommand.

etungsten committed:
pubsys: add `check-repo-expirations` to check metadata expirations

Adds new pubsys subcommand `check-repo-expirations` that checks for
repository metadata expirations within a specified timeframe

Adds a new Makefile.toml task that invokes `check-repo-expirations`

etungsten commented:
Push above does several things to address https://github.com/bottlerocket-os/bottlerocket/pull/1130#discussion_r497185358:

- All new subcommands take `--arch` and `--variant` as arguments for dealing with a single repository
- All new Makefile.toml tasks calls the subcommands with `BUILDSYS_ARCH` and `BUILDSYS_VARIANT` passed to `--arch` and `--variant` respectively.
- All new subcommands can now also take a `--repo-manifest` argument that is exclusive from `--arch`, `--variant` which can be used to specify a .toml file that contains set of repositories created using the same named repo infrastructure from Infra.toml. The expected format looks like the following:
```
arches = [ "x86_64", "aarch64" ]
variants = [
   "aws-k8s-1.15",
   "aws-k8s-1.16",
   "aws-k8s-1.17",
   "aws-ecs-1",
]
```

Tested all the subcommands and different argument combinations. Updating PR description with new testing.

etungsten commented:
Push above drops an extraneous commit I did not intend to include here.

etungsten commented:
Push above reduces level of indentation and some minor comment fixes.

etungsten commented:
Push above removes and refactors code for processing a manifest of repositories. Now all new subcommands only take `--arch` and `--variant` for processing a single TUF repository.

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+pub(crate) struct CheckExpirationsArgs {
+    #[structopt(long)]
+    /// Use this named set of repositories from Infra.toml
```
Below we say `Use this named repo from Infra.toml` and `Use this named repo infrastructure from Infra.toml`, and given the confusion about single TUF repo vs. set of TUF repos we're having, maybe it'd be worthwhile to consolidate on one wording...

tjkirch replied:
It looks like all of the repo args except for this one were updated :)

tjkirch replied:
It looks like all of the repo args except for this one were updated :)

tjkirch commented on Makefile.toml:
```diff
@@ -340,6 +346,22 @@ pubsys-setup \
 '''
 ]
 
+[tasks.publish-setup-without-key]
```
I'm not sure this task is worth it.  I think we'll generally always specify a role and key.  We're not saying we're going to use the key, we just need it defined.  There will always be a root key if there's a root role (I think we can ignore unsigned root roles) so it seems like we're talking about a case of someone knowing the root role and wanting to verify it but not being the keyholder, for example people outside of the Bottlerocket project wanting to validate the Bottlerocket repos.  (It's similar to listing Bottlerocket repos; they couldn't sign a repo they started from the official ones.)  It seems to me like that's safely in the owner's domain.  Am I missing a use case where it's important to validate without even having the ability to specify a key?  (You could even list file:///dev/null...)

If we want that capability, I'd rather do it without duplicating the task here.  I think we can do it with cargo-make, like this example from their docs:
```
[tasks.test-flow]
env = { "SOME_ENV_VAR" = "value" }
run_task = "actual-task"
```
in our case:
```
[tasks.publish-setup-without-key]
env = { "ALLOW_MISSING_KEY_ARG" = "true" }
run_task = "publish-setup"
```

etungsten replied:
I chose to keep the `publish-setup-without-key` task. Regardless of intent, I think being able to load a repository and check expirations shouldn't need the repo signing key specified.

tjkirch commented on Makefile.toml:
```diff
@@ -407,6 +429,77 @@ ln -sfn "${PUBLISH_REPO_OUTPUT_DIR##*/}" "${PUBLISH_REPO_OUTPUT_DIR%/*}/latest"
+   --variant "${BUILDSYS_VARIANT}" \
+   \
+   --root-role-path "${PUBLISH_REPO_ROOT_JSON}" \
+   --expirations-between-now-and "in ${EXPIRING_WITHIN}"
```
```suggestion
   --expiring-within "${EXPIRING_WITHIN}"
```
(`parse_datetime` doesn't require the "in")

(I think "expirations between now and" is OK within the code, because the `parse_datetime` call effectively transforms the meaning of the variable.  "expiration limit" is also good, which you use later in the code.  It'd be good to standardize on one within the Rust code, though.)

tjkirch commented on Makefile.toml:
```diff
@@ -407,6 +429,77 @@ ln -sfn "${PUBLISH_REPO_OUTPUT_DIR##*/}" "${PUBLISH_REPO_OUTPUT_DIR%/*}/latest"
+   \
+   --root-role-path "${PUBLISH_REPO_ROOT_JSON}" \
+   --repo-expiration-policy-path "${PUBLISH_EXPIRATION_POLICY_PATH}" \
+   ${UNSAFE_REFRESH:+--unsafe-refresh} \
```
I'd recommend doing the same thing here that we did for `ALLOW_MISSING_KEY` to make sure it's set to "true", otherwise people could say `UNSAFE_REFRESH=false` to try and be safe, and get the wrong result.

tjkirch commented on tools/pubsys/src/main.rs:
```diff
@@ -138,6 +156,24 @@ mod error {
+            source: crate::repo::validate_repo::Error,
+        },
+
+        #[snafu(context(false), display("{}", source))]
```
I notice that all the new variants refuse context, but two of three have a `display` attribute with text, and this one has an empty one.  Do you need a `display` attribute at all with `context(false)`?  `context(false)` says there's no useful information to add.  I think it might be better to remove that and have a real display attribute for all three, given that the inner errors might be clearer with the context that you were trying to refresh a repo, for example.

etungsten replied:
You don't need a `display` attribute with `context(false)` but what ends up happening is the enum variant is used as the context. So in this case, the error string will look like "CheckExpirations: Failed to blah blah blah". 

I'll go ahead remove the `context(false)` and specify context for all three error variants.

tjkirch replied:
Should we do the same for the `Repo` variants in the three new modules?

etungsten replied:
The `Repo` variants are as intended, I have no additional context to add.

tjkirch commented on tools/pubsys/src/repo.rs:
```diff
@@ -588,6 +636,9 @@ mod error {
         },
 
+        #[snafu(display("Repo configuration is invalid"))]
+        InvalidRepoConfig,
```
This doesn't give the user any context on what was wrong with their config.  Currently it could only mean that an invalid metadata URL was given, but it'd be good to include some context.

It seems like we don't need this error variant here, because it's only used for return values of `repo_urls`, which is in this module and already wraps its errors in this module's error type.  How about in the places where we call `repo_urls`, we do `.context(error::RepoUrlsError)?`, where `RepoUrlsError` has a `source: repo::Error`?

[edit] I think I misunderstood; I notice that we're `?`-ing the `Result` of `repo_urls` and `context`-ing the inner `Option` of `repo_urls`, which I think will pass on the `repo` error if the file is invalid, and use the `InvalidRepoConfig` for `None` which implies they didn't specify URLs.  If that's correct, can we clarify this message to say that they didn't specify repo URLs for repo X?

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+//! The check_expirations module owns the 'check-repo-expirations' subcommand.
```
Can we extend the docstrings a bit and explain the purpose of the module/subcommand?

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    expirations_between_now_and: Option<DateTime<Utc>>,
+}
+
+/// Checks for upcoming role expirations, gathers them in a list along with their expiration date.
```
```suggestion
/// Checks for upcoming role expirations, gathering them in a map of role to expiration datetime.
```

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    info!("Snapshot expiration:\t{}", repo.snapshot().signed.expires);
+    info!("Targets expiration:\t{}", repo.targets().signed.expires);
+    info!("Timestamp expiration:\t{}", repo.timestamp().signed.expires);
+    // Check for upcoming metadata expirations if a timeframe is specified
+    if let Some(expiration_limit) = expirations_between_now_and {
```
What's the goal of calling this subcommand without an expiration time?  Can we get rid of the Option?

etungsten replied:
If this subcommand is called without an expiration time. It just checks for expirations as of right now. Which is functionally the same thing as `validate-repo` but without downloading any targets.

etungsten replied:
Well not exactly the same. `validate-repo` won't list out all the expired roles and just the first one where as in `check-repo-expirations` it will list out all of them.

tjkirch replied:
We have a default in the makefile, though.  Someone would have to call pubsys manually without a date to hit this case, right?  I don't think that's worth it..?

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    if let Some(expiration_limit) = expirations_between_now_and {
+        let upcoming_expirations = find_upcoming_metadata_expiration(&repo, expiration_limit);
+        if !upcoming_expirations.is_empty() {
+            for (role, date) in upcoming_expirations {
```
```suggestion
            for (role, expiration_date) in upcoming_expirations {
```

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+                } else {
+                    warn!(
+                        "Repo '{}': '{}' expiring between now and {} on {}",
+                        metadata_url, role, expiration_limit, date
```
```suggestion
                        "Repo '{}': '{}' expiring in {} at {}",
                        metadata_url, role, date - Utc::now(), date
```
(maybe a bit clearer; might want to save one Utc::now() before the loop)

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+        })?
+        .get(&check_expirations_args.repo)
+        .context(repo_error::MissingConfig {
+            missing: format!("definition for repo {}", &check_expirations_args.repo),
```
```suggestion
        .with_context(|| repo_error::MissingConfig {
            missing: format!("definition for repo {}", &check_expirations_args.repo),
```

tjkirch replied:
I noticed a few other `context` calls change to `with_context` even though there's no copying/`format` inside, is there a benefit to that?  The goal of this suggestion was preventing a string build in the non-error case, since it was happening outside the closure.

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+        })?;
+
+    let transport = RepoTransport::default();
+    // If we get both the variant and arch specified, just check expirations for that specific repository
```
```suggestion
```
(I think that's the only option now)

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility = "pub(super)")]
+    pub(crate) enum Error {
+        #[snafu(context(false), display("{}", source))]
+        Repo { source: crate::repo::Error },
+
```
```suggestion
```
Doesn't seem used?  Same in other modules.  (Though, see my other suggestion about using it for repo_urls...)

etungsten replied:
This is used for `?`ing directly on `crate::repo::error::Error`s. Maybe I should go ahead and add context to those errors. The only problem is that I started noticing very very verbose error output and I fear adding another redundant level of nesting.

tjkirch replied:
Yeah, if it's just repeating "Tried a thing: tried a thing: tried a thing: thing failed", then that's not helpful.  We do have that in a couple spots.  I wasn't expecting it here, but if that's the case, no problem!

tjkirch commented on tools/pubsys/src/repo/check_expirations/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    #[structopt(long, requires = "variant")]
+    /// The architecture of the repo being checked for expirations
+    arch: String,
+    #[structopt(long, requires = "arch")]
```
I'm not sure I understand the `requires` here, and in the other modules; aren't they both required?  (For now)

tjkirch commented on tools/pubsys/src/repo/refresh_repo/mod.rs:
```diff
@@ -0,0 +1,210 @@
+    /// expired metadata files.
+    ///
+    /// CAUTION: TUF metadata expiration dates are designed to prevent freeze attacks.
+    /// By setting this flag, you are bypassing this security feature of TUF.
```
I'm not sure the owner of the repo, the one refreshing it, is liable to a freeze attack in this use case?  The expiration dates are almost meaningless if we're replacing them.

tjkirch commented on tools/pubsys/src/repo/refresh_repo/mod.rs:
```diff
@@ -0,0 +1,210 @@
+    set_versions(&mut repo_editor)?;
+
+    // Get the signing keys and sign the repository
+    let key_source = get_signing_key_source(signing_key_config);
```
Should we pass the `key_source` directly to `refresh_repo` since we're not using anything else from `signing_key_config`?  Or do you anticipate other needs?

etungsten replied:
Oops. I missed this.

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,228 @@
+//! The validate_repo module owns the 'validate-ami' subcommand and provides methods for validating
+//! a given set of TUF repositories by downloading its targets and check for metadata expirations.
```
```suggestion
//! a given set of TUF repositories by downloading its targets and checking for metadata expirations.
```

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,228 @@
+    info!("Loaded TUF repo: {}", metadata_url);
+    // Try retrieving listed targets
+    let percentage = if percent_target_files > 100 {
+        100
```
Aww, don't let them get away with that nonsense ;)  Hard fail via `ensure`!

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,228 @@
+        .collect();
+
+    for target in sampled_targets {
+        let maybe_target_reader = attempt_read_target(repo, target.to_string())?;
```
Validation can take quite a while, and it's a mix of CPU and IO.  It seems like it'd be worth it to do a few checks in parallel?  We have some examples of that in pubsys, e.g. setting SSM parameters in parallel, or describing AMIs.  Could just be a TODO, but I think it's worth it to speed this up if possible.

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,228 @@
+                        "Repo target fetch failure on attempt {} out of {}",
+                        attempt, max_attempts
+                    );
+                    if attempt >= max_attempts {
+                        return Err(repo_error::Error::ReadTarget {
```
```suggestion
                    ensure!(attempt < max_attempts, repo_error::Error::ReadTarget {
```
(same in download)

etungsten replied:
I think the issue here is that I can't use the enum variant directly with `ensure!` where I would be able to specify the `source` field. I want to return `error::Error::ReadTarget` where source is set to `err`.

`ensure!` forces me to use snafu's context selector type (e.g. error::ReadTarget).

```
error[E0560]: struct `repo::error::ReadTarget<std::string::String>` has no field named `source`
  --> pubsys/src/repo/validate_repo/mod.rs:65:29
   |
65 | ...                   source: err
   |                       ^^^^^^ `repo::error::ReadTarget<_>` does not have this field
   |
   = note: available fields are: `target`

error[E0599]: no method named `fail` found for struct `repo::error::ReadTarget<std::string::String>` in the current scope
   --> pubsys/src/repo/validate_repo/mod.rs:61:21
    |
61  | /                     ensure!(
62  | |                         attempt < max_attempts,
63  | |                         repo_error::ReadTarget {
64  | |                             target,
65  | |                             source: err
66  | |                         }
67  | |                     );
    | |______________________^ method not found in `repo::error::ReadTarget<std::string::String>`
    | 
   ::: pubsys/src/repo.rs:591:21
    |
591 |       #[derive(Debug, Snafu)]
    |                       ----- method `fail` not found for this
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,228 @@
+    let max_attempts = 5;
+    for attempt in 1..=max_attempts {
+        info!("Downloading target: {}", target);
+        if let Err(err) = io::copy(&mut reader, &mut io::sink()) {
```
I think it's worth a comment here explaining that tough's reader implementation is validating the target as it's pulled, so we're not just wasting bandwidth here.

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,228 @@
+    let percentage = f32::from(percentage) / 100.0;
+    let num_to_retrieve = (targets.len() as f32 * percentage).ceil();
+    let mut rng = &mut rand::thread_rng();
+    let mut sampled_targets: Vec<String> = targets.keys().map(|key| key.to_string()).collect();
+    sampled_targets = sampled_targets
+        .choose_multiple(&mut rng, num_to_retrieve as usize)
+        .cloned()
+        .collect();
```
I'm not sure we need to clone twice here; I was able to get this to work with a `HashMap<String, foo>` like we get from `targets().signed.targets`:
```
    let keys = map.keys().collect::<Vec<&String>>();
    for selected in keys.choose_multiple(&mut rng, 1) {
```
Would that work here?  We're calling `to_string` / `clone` below anyway, where we need ownership...

etungsten replied:
After switching to using async, the compiler keeps complaining about the lifetime of `targets` being not long enough. So I ended up still having to clone the keys and the sampled targets.

tjkirch replied:
I think it might just need the `.choose_multiple.cloned`, and not the prior `String`/`cloned` when defining `targets` because each `target` is all that goes into the async block.  (Alternatively, maybe `let target_str = target.to_string();` before the `async move`, and using `target_str` inside the `async move`, would be enough to clarify that only a single owned value is moving in.)

This is all just a performance nit, we can move on if it's being difficult!

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,228 @@
+
+    for target in sampled_targets {
+        let maybe_target_reader = attempt_read_target(repo, target.to_string())?;
+        let reader = maybe_target_reader.context(error::TargetMissing {
```
Do we need the separate variables rather than calling `.context` directly on the result of `attempt_read_target`?  I think that's more standard, unless you're doing something special with the Err case.

tjkirch reviewed (commented):
Nice job, this will be useful!

zmrow reviewed (approved):
Provided @tjkirch 's comments are addressed - I'm happy. 

:jack_o_lantern:

etungsten commented:
Addressed @tjkirch 's review comments from https://github.com/bottlerocket-os/bottlerocket/pull/1130#pullrequestreview-505165582.

Big changes:
- Validate repo now spawns threads to download and validate targets in parallel (not sure if this is idiomatic, please check my work!)

Seems to work:
```
$ cargo make -e PUBLISH_REPO="2020-07-07" -e BUILDSYS_VARIANT="aws-k8s-1.17" -e BUILDSYS_ARCH="x86_64" validate-repo
[cargo-make] INFO - cargo make 0.32.7
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: validate-repo
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup
18:11:28 [INFO] Found infra config at path: Infra.toml
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: validate-repo
18:11:29 [INFO] Using infra config from path: Infra.toml
18:11:29 [INFO] Loaded TUF repo: https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/x86_64
18:11:29 [INFO] Attempting to download 100% of listed targets from https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/x86_64 for validation purposes
18:11:29 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.0-b0e2bc22-root.verity.lz4
18:11:29 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-0.5.0-e0ddf1b-boot.ext4.lz4
18:11:30 [INFO] Downloading target: migrate_v1.0.0_ecr-helper-admin.lz4
18:11:30 [INFO] Downloading target: migrate_v0.4.1_pivot-repo-2020-07-07.lz4
18:11:30 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-0.4.1-a29a1450-boot.ext4.lz4
18:11:30 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.1-2a181156-root.ext4.lz4
18:11:30 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.0-b0e2bc22-root.ext4.lz4
18:11:30 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-0.4.1-a29a1450-root.ext4.lz4
18:11:30 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.0-b0e2bc22-boot.ext4.lz4
18:11:30 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.1-2a181156-boot.ext4.lz4
18:11:31 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-0.4.1-a29a1450-root.verity.lz4
18:11:31 [INFO] Downloading target: migrate_v1.0.0_ecr-helper-control.lz4
18:11:31 [INFO] Downloading target: migrate_v0.4.1_add-version-lock-ignore-waves.lz4
18:11:31 [INFO] Downloading target: migrate_v0.3.2_admin-container-v0-5-0.lz4
18:11:32 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-0.5.0-e0ddf1b-root.verity.lz4
18:11:32 [INFO] Downloading target: manifest.json
18:11:32 [INFO] Downloading target: migrate_v0.5.0_control-container-v0-4-1.lz4
18:11:37 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-0.5.0-e0ddf1b-root.ext4.lz4
18:11:37 [INFO] Downloading target: bottlerocket-aws-k8s-1.17-x86_64-1.0.1-2a181156-root.verity.lz4
18:11:37 [INFO] Downloading target: migrate_v0.5.0_admin-container-v0-5-2.lz4
18:11:42 [INFO] Downloading target: migrate_v0.5.0_add-cluster-domain.lz4
[cargo-make] INFO - Build Done in 32 seconds.

```

etungsten commented:
Push above rebases onto develop to include a fix for the CI action runners.

webern reviewed (approved):
async usage lgtm

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,252 @@
+//! The validate_repo module owns the 'validate-ami' subcommand and provides methods for validating
+//! a given TUF repository by attempting to loadi the repository and downloading its targets.
```
```suggestion
//! a given TUF repository by attempting to load the repository and download its targets.
```

tjkirch commented on tools/pubsys/src/repo/refresh_repo/mod.rs:
```diff
@@ -0,0 +1,210 @@
+//! The refresh_repo module owns the 'refresh-repo-metadata' subcommand and provide methods for
+//! refreshing and resigning the metadata files of a given TUF repository.
```
```suggestion
//! refreshing and re-signing the metadata files of a given TUF repository.
```

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,252 @@
+    Ok(None)
+}
+
+async fn attempt_download_target(
```
Hmm, why does this need to be async now?  The implementation didn't change, and it seems synchronous.

etungsten replied:
You're right, this doesn't need to be async. Removing..

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,252 @@
+        #[snafu(display("Missing target: {}", target))]
+        TargetMissing { target: String },
+
+        #[snafu(display("Failed to spawn threads for fetching target: {}", source))]
```
```suggestion
        #[snafu(display("Failed to spawn task for fetching target: {}", source))]
```
(tokio decides whether a task will be in a new thread or the current one)

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,252 @@
+        .choose_multiple(&mut rng, num_to_retrieve as usize)
+        .cloned()
+    {
+        let reader =
+            attempt_read_target(&repo, target.to_string())?.context(error::TargetMissing {
+                target: target.clone(),
+            })?;
+
+        tasks.push(tokio::spawn(async move {
+            attempt_download_target(reader, target.to_string()).await
+        }));
```
I think we may want to do both `attempt_read_target` and `attempt_download_target` in the spawned task, so we can add delays between retry attempts in both; it's not as effective to try immediately, we could wait a second.

(I wonder if the functions should be combined, actually.  If a download fails, reusing the same reader object seems dangerous, right?  `Read`ers don't seek back to the start, so it'd only work if the client handled range requests, I think?  Not all `Read`ers would still be valid after a failure.)

etungsten replied:
I think tough's `HttpTransport` handles retries and sends range request as a consequence of @webern 's work in https://github.com/awslabs/tough/pull/143.

So actually I'm not sure if it's worth implementing the retries here at all...

tjkirch replied:
I'd be OK with removing it; as long as it's going through the transport, that's a better spot for the logic.  If you'd prefer not to, I think it'd make sense to combine read/download, so we don't reuse readers.

etungsten commented:
Push above addresses @tjkirch 's remaining comments.

Push below fixes a minor ownership issue.

tjkirch commented on tools/pubsys/src/repo/refresh_repo/mod.rs:
```diff
@@ -0,0 +1,210 @@
+    }
+}
+pub(crate) use error::Error;
+use tough::key_source::KeySource;
```
nit: can this be up with the other `use`s?  :)

etungsten commented:
Push above removes the retries for fetching and downloading targets. tough's `HttpTransport` already implements retries.
Tested things and they still work as before..

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,191 @@
+        info!("Downloading target: {}", target);
+        tasks.push(tokio::spawn(async move {
+            // tough's `Read` implementation validates the target as it's being downloaded
+            io::copy(&mut reader, &mut io::sink()).context(error::TargetDownload)
```
I think we can include `target` in the error context since we have ownership here, right?

tjkirch commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,191 @@
+        let target = target.to_string();
+        let mut reader = repo
+            .read_target(&target)
+            .context(repo_error::ReadTarget {
+                target: target.to_string(),
+            })?
+            .context(error::TargetMissing {
```
```suggestion
            .with_context(|| repo_error::ReadTarget {
                target: target.to_string(),
            })?
            .with_context(|| error::TargetMissing {
```

etungsten replied:
I think I only need to use `with_context` for `TargetMissing` right? To convert the `Option` into a `Result`?

etungsten replied:
Oh nvm, I understand now. We're doing this to prevent a string build in non-error cases.

bcressey commented on Makefile.toml:
```diff
@@ -42,6 +42,12 @@ PUBLISH_DATA_VOLUME_SIZE = "20"
 # tools/pubsys/policies/ssm/README.md for more information.
 PUBLISH_SSM_TEMPLATES_PATH = "${BUILDSYS_ROOT_DIR}/tools/pubsys/policies/ssm/defaults.toml"
 
+# Percentage of TUF repository targets to sample and retrieve as part of the repo validation process
+VALIDATE_PERCENT_TARGETS = "100"
+# Specifies the timeframe to look for upcoming repository metadata expirations
+EXPIRING_WITHIN = "3 days"
+# When refreshing repositories, you can set UNSAFE_REFRESH=true to refresh repositories that have expired metadata files.
```
nit: I'd prefer that these are all prefixed in a consistent way, e.g. `VALIDATE_`, so it's easier to map to the eventual use case, and so we know we're not setting a variable that has meaning elsewhere.

tjkirch replied:
VALIDATE_EXPIRATIONS_WITHIN?

etungsten replied:
Right now the `check-repo-expiration` task is somewhat functionally separate from the `validate-repo` task. I don't want to give the impression that `check-repo-expiration` should be used to validate a repository (it only checks for metadata expirations and that's it).

Perhaps METADATA_EXPIRING_WITHIN? Would that be more clear?

etungsten replied:
Actually maybe a `REPO_` prefix would be better...

bcressey commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,193 @@
+//! The validate_repo module owns the 'validate-ami' subcommand and provides methods for validating
```
```suggestion
//! The validate_repo module owns the 'validate-repo' subcommand and provides methods for validating
```

bcressey commented on tools/pubsys/src/repo/refresh_repo/mod.rs:
```diff
@@ -0,0 +1,210 @@
+//! The refresh_repo module owns the 'refresh-repo-metadata' subcommand and provide methods for
```
```suggestion
//! The refresh_repo module owns the 'refresh-repo' subcommand and provide methods for
```

bcressey commented on tools/pubsys/src/repo/validate_repo/mod.rs:
```diff
@@ -0,0 +1,193 @@
+        metadata_base_url: metadata_url.clone(),
+    })?;
+    info!("Loaded TUF repo: {}", metadata_url);
+    // Try retrieving listed targets
+    info!(
+        "Attempting to download {}% of listed targets from {} for validation purposes",
+        percent_target_files, metadata_url
+    );
+    retrieve_percentage_of_targets(&repo, percent_target_files).await?;
```
It's hard for me to reason about why we need this or what percentages other than 0% or 100% would make sense.

The advantage of the random subset is that we can converge on full validation over time without paying the cost during each run, but the drawback is that it's hard to know precisely when that will happen. It also doesn't help us put a time bound on execution since some of the targets are much larger than others (migration binary vs. root filesystem image).

What do you think about having a modal behavior, either "deep" (100%) or "shallow" (0%)? Then shallow checks could be scheduled more frequently and deep checks could run hourly or daily.

etungsten replied:
Yeah, ideally the targets sampling would have worked more deterministically to cover all targets over some number of runs. But to implement something like that would be pretty complex and probably more trouble than it's worth.

I'll go ahead and change `--percent-target-files` to an optional flag `--validate-targets` to download all targets. By default we would just load the repository and make sure we can list the targets.

I think maybe a sensible addition would be to target files for specific versions of Bottlerocket (images, migrations) for validation but that is probably out of the scope of this PR.

tjkirch replied:
I think we'll have to figure out a solution for validating targets without linearly increasing the time between metrics as we have more releases.  (Even percentage wouldn't do that.)  If we had a way of distinguishing new versus old targets (for us, the version numbers in the filenames?) then we could validate newer targets more frequently and do full runs as a batch job, perhaps.

(Not as a blocker here, of course)

etungsten replied:
I will create an issue to track this (have the ability to validate "newer" targets) once this PR is merged.

etungsten replied:
https://github.com/bottlerocket-os/bottlerocket/issues/1172

etungsten commented:
Addresses @bcressey 's comments.

- Added `REPO_` prefix to cargo-make env vars related to *-repo tasks.
- Removed random target sampling and added a flag to denote whether to download all targets to validate or not

Tested changes and they work as expected.
